### PR TITLE
feat(ecs): deployment circuit breaker support

### PR DESCRIPTION
**Pull request by pahud**
feat(ecs): deployment circuit breaker support

This PR allows you to enable the `deployment circuit breaker` support for Amazon ECS.

Depends on 

- [ ] cfnspec v22 https://github.com/aws/aws-cdk/issues/12170

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by pahud** [comment]
copy @iamhopaul123

File: packages/@aws-cdk/aws-ecs/README.md [hunk 1]
```diff
@@ -335,6 +335,35 @@ const service = new ecs.FargateService(this, 'Service', {
 `Services` by default will create a security group if not provided.
 If you'd like to specify which security groups to use you can override the `securityGroups` property.
 
+### Deployment circuit breaker and rollback
+
+By default, Amazon ECS [deployment circuit breaker](https://aws.amazon.com/tw/blogs/containers/announcing-amazon-ecs-deployment-circuit-breaker/)
+is enabled and will rollback on deployment failure. With this feature, Amazon ECS will automatically roll back unhealthy
+service deployments without the need for manual intervention. As this feature with AWS Fargate requires
+platform `1.4.0` or above. Make sure to specify the `platformVersion` to `1.4` as the current default version points to `1.3`.
```
Comment (tobli) [inline packages/@aws-cdk/aws-ecs/README.md:343]: I don't think this is correct, we've tried circuit breaker with platform version 1.3. The documentation doesn't mention this being a requirement.

Comment (pahud) [reply packages/@aws-cdk/aws-ecs/README.md:343]: > Is this correct? The documentation doesn't mention this being a requirement afaik.

Thanks for bringing it up. I'll confirm this internally.

**Review by iamhopaul123 — COMMENTED** [review]
Thank you so much for contributing! Overall looks good to me. Just one suggestion.

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts [hunk 2]
```diff
@@ -161,6 +161,19 @@ export interface BaseServiceOptions {
...
+   * @default false
+   */
+  readonly deploymentCircuitBreaker?: boolean;
+
+  /**
+   * Whether to enable Amazon ECS to roll back the service if a service deployment fails. If rollback is enabled,
+   * when a service deployment fails, the service is rolled back to the last deployment that completed successfully.
+   * @default false
+   */
+  readonly deploymentRollback?: boolean;
```
[the remark this answers is not in the document]
Comment (pahud) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:176]: Agree. Let's do it.

[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:176]: Sounds good! Before merging in, could you manually test if this feature works and if it doesn't work for `enable=false` and `rollback=true`?

[the remark this answers is not in the document]
Comment (pahud) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: Hi @iamhopaul123 

I just tried this

```ts
const service = new ecs.FargateService(stack, 'Service', {
  cluster,
  desiredCount: 5,
  taskDefinition,
  circuitBreaker: { rollback: false },
});
```

圖片

The `PRIMARY` deployment `rolloutState` remained `IN_PROGRESS` for a long time with no rollback with some `failedTasks`.  It seemed to be working as expected ( circuit breaker with no rollback). However, the number of `failedTasks` was increasing very slowly.


Then I tried 

```ts
const service = new ecs.FargateService(stack, 'Service', {
  cluster,
  desiredCount: 5,
  taskDefinition,
  circuitBreaker: { rollback: true },
});
```

Everything worked as expected.

圖片

**Review by iamhopaul123 — DISMISSED** [review]
Awesome! Just got confirmed internally that `rollback` depends on the CB and doesn't make any sense if CB is not enabled, though it won't error out because it doesn't validate deployment configuration.

**Comment by iamhopaul123** [comment]
@MrArnoldPalmer any thought or concern on this?

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts [hunk 3]
```diff
@@ -161,6 +173,12 @@ export interface BaseServiceOptions {
    * @default - Rolling update (ECS)
    */
   readonly deploymentController?: DeploymentController;
+
+  /**
+   * Whether to enable the deployment circuit breaker
+   * @default - disabled
+   */
+  readonly circuitBreaker?: DeploymentCircuitBreaker;
```
Comment (SoManyHs) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:181]: I see in line 380 that we are always setting "enable" to be true, which makes this comment about the default setting a little confusing. 

If the `DeploymentCircuitBreaker` only contains the boolean `rollback`, would it make sense then to just have this is a boolean field on `BaseServiceOptions`?

**Review by SoManyHs — DISMISSED** [review]

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts (hunk 3 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:181]: @SoManyHs  OK Let me do it now.

Comment (pahud) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:182]: Just updated the description. Let me know if it looks better now.

@SoManyHs

Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:182]: @SoManyHs as for now it just has `rollback` but yes potentially in the future they'll add more options to it.

**Review by SoManyHs — APPROVED** [review]

**Pull request merged**